{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7add1629",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pandas import read_csv\n",
    "import requests\n",
    "import json\n",
    "import math\n",
    "import time\n",
    "from datetime import datetime\n",
    "import copy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cab9f53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove dictionary keys for empty lists and null values\n",
    "def cleanNullTerms(d):\n",
    "    clean = {}   \n",
    "    for k, v in d.items():\n",
    "        if v is not None and len(v)>0:\n",
    "            clean[k] = v   \n",
    "    return clean\n",
    "\n",
    "## Fetch the data from biotools\n",
    "def get_pages(biotoolsapiurl):\n",
    "    biotoolsapiurl\n",
    "    payloads = {'format': 'json', 'domain': 'covid-19','page':'i'}\n",
    "    r = requests.get(biotoolsapiurl, params=payloads, timeout=20).json()\n",
    "    count = r['count']\n",
    "    list_num = len(r['list'])\n",
    "    total_pages = math.ceil(count/list_num)\n",
    "    return total_pages\n",
    "\n",
    "## Pull the key for dictionaries with only one key-value pair\n",
    "def get_dict_key(single_dict):\n",
    "    keylist = list(single_dict.keys())\n",
    "    only_key = keylist[0]\n",
    "    return only_key\n",
    "\n",
    "## Parse defined terms and format them as a Defined Term object\n",
    "def parse_defined_terms(definedtermlist):\n",
    "    try:\n",
    "        termdf = pd.DataFrame(definedtermlist)\n",
    "        termdf.rename(columns={'uri':'url','term':'name'},inplace=True)\n",
    "        termdf['inDefinedTermSet']= 'http://edamontology.org/'\n",
    "        termdf['@id'] = termdf['url'].str.replace('http://edamontology.org/','')\n",
    "        termdf['@type'] = 'schema:DefinedTerm'\n",
    "        termdf['termCode'] = termdf['@id']\n",
    "        cleandf = termdf[['@type','@id','inDefinedTermSet','termCode','url','name']]\n",
    "        termjson = cleandf.to_dict(orient=\"records\")\n",
    "    except:\n",
    "        termjson = -1\n",
    "    return termjson\n",
    "\n",
    "## Parse input and outputs and format as a list of Formal Parameter objects\n",
    "def parse_parameters(parameterdict):\n",
    "    datalist = []\n",
    "    formatlist = []\n",
    "    for eachparameter in parameterdict:\n",
    "        content_types = list(eachparameter.keys())\n",
    "        if 'data' in content_types:\n",
    "            if isinstance(eachparameter['data'], dict) == True:\n",
    "                datadictlist = []\n",
    "                datadictlist.append(eachparameter['data'])\n",
    "            elif isinstance(eachparameter['data'], list) == True:\n",
    "                datadictlist = eachparameter['data']\n",
    "            datavalue = parse_defined_terms(datadictlist)\n",
    "            if datavalue != -1:\n",
    "                datalist.extend(datavalue)               \n",
    "        if 'format' in content_types:\n",
    "            if isinstance(eachparameter['format'], dict) == True:\n",
    "                formatdictlist = []\n",
    "                formatdictlist.append(eachparameter['format'])\n",
    "            elif isinstance(eachparameter['format'], list) == True:\n",
    "                formatdictlist = eachparameter['format']\n",
    "            formatvalue = parse_defined_terms(formatdictlist)\n",
    "            if formatvalue != -1:\n",
    "                formatlist.extend(formatvalue)\n",
    "    if len(datalist) > 0 or len(formatlist) > 0:\n",
    "        parameterjson = {'@type':'bioschemastypes:FormalParameter',\n",
    "                         'encodingFormat': datalist,\n",
    "                         'defaultValue': formatlist}\n",
    "    else:\n",
    "        parameterjson = -1\n",
    "    return parameterjson\n",
    "\n",
    "## Create an outbreak-compatible json\n",
    "class cleandoc:\n",
    "    now = datetime.now()\n",
    "    basejson = {\n",
    "      \"@type\": \"outbreak:ComputationalTool\",\n",
    "      \"@context\": {\n",
    "        \"schema\": \"http://schema.org/\",\n",
    "        \"outbreak\": \"https://discovery.biothings.io/view/outbreak/\",\n",
    "        \"dct\": \"http://purl.org/dc/terms/\"\n",
    "      },\n",
    "      \"curatedBy\": {\n",
    "        \"@type\": \"outbreak:Organization\",\n",
    "        \"name\": \"ELIXIR bio.tools\",\n",
    "        \"url\": \"https://bio.tools/t?domain=covid-19\",\n",
    "        \"identifier\": \"biotools\",\n",
    "        \"affiliation\": [{\"@type\":\"outbreak:Organization\",\"name\": \"ELIXIR\"}],\n",
    "        \"curationDate\": now.strftime(\"%Y-%m-%d\")\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    ## Directly add any key/value pair that has minimal transformations\n",
    "    def add_basic_info(biotooljsonhit): \n",
    "        cleanjson = cleandoc.basejson\n",
    "        cleanjson['name'] = biotooljsonhit['name']\n",
    "        cleanjson['description'] = biotooljsonhit['description']\n",
    "        cleanjson['identifier'] = biotooljsonhit['biotoolsID']\n",
    "        cleanjson['url'] = biotooljsonhit['homepage']\n",
    "        cleanjson['softwareVersion'] = biotooljsonhit['version']\n",
    "        cleanjson['applicationCategory'] = biotooljsonhit['toolType']\n",
    "        cleanjson['license'] = biotooljsonhit['license']\n",
    "        cleanjson['programmingLanguage'] = biotooljsonhit['language']\n",
    "        if biotooljsonhit['accessibility'] != None:\n",
    "            cleanjson['conditionsOfAccess'] = biotooljsonhit['accessibility']\n",
    "        elif biotooljsonhit['cost'] != None:\n",
    "            cleanjson['conditionsOfAccess'] = biotooljsonhit['cost']\n",
    "        try:\n",
    "            cleanjson['dateModified'] = biotooljsonhit['lastUpdate'].split('T')[0]\n",
    "        except:\n",
    "            cleanjson['dateModified'] = biotooljsonhit['lastUpdate']\n",
    "        try:\n",
    "            cleanjson['dateCreated'] = biotooljsonhit['additionDate'].split('T')[0]\n",
    "        except:\n",
    "            cleanjson['dateCreated'] = biotooljsonhit['additionDate']          \n",
    "        return cleanjson\n",
    "\n",
    "    ## Add application subcategory after transforming it\n",
    "    def add_app_sub_cat(cleanjson,biotooljsonhit): \n",
    "        try:\n",
    "            alltopics = biotooljsonhit['topic']\n",
    "            topicjson = parse_defined_terms(alltopics)\n",
    "            if topicjson != -1:\n",
    "                cleanjson['applicationSubCategory'] = topicjson\n",
    "        except:\n",
    "            pass\n",
    "        return cleanjson\n",
    "    \n",
    "    ## Add featuress after transforming them        \n",
    "    def add_features(cleanjson,biotooljsonhit):\n",
    "        available_functions = biotooljsonhit['function']\n",
    "        if len(available_functions) > 0:\n",
    "            operationlist = []\n",
    "            inputlist = []\n",
    "            outputlist = []\n",
    "        for eachfunction in available_functions:\n",
    "            ## parse operations\n",
    "            operations = eachfunction['operation']\n",
    "            features = parse_defined_terms(operations)\n",
    "            if features != -1:\n",
    "                operationlist.extend(features)\n",
    "            ## parse inputs\n",
    "            inparameterdict = eachfunction['input']\n",
    "            inparameters = parse_parameters(inparameterdict)\n",
    "            if inparameters != -1:\n",
    "                inputlist.append(inparameters)\n",
    "            ## parse outputs\n",
    "            outparameterdict = eachfunction['output']\n",
    "            outparameters = parse_parameters(outparameterdict)\n",
    "            if outparameters != -1:\n",
    "                outputlist.append(outparameters)\n",
    "                ## Note, additional content available from bio.tools in this section includes 'cmd', and 'notes'\n",
    "            if len(operationlist)>0:\n",
    "                cleanjson['featureList'] = operationlist\n",
    "            if len(inputlist)>0:\n",
    "                cleanjson['input'] = inputlist\n",
    "            if len(outputlist)>0:\n",
    "                cleanjson['output'] = outputlist\n",
    "        return cleanjson\n",
    "\n",
    "    ## get the downloadUrl and codeRepository from either link or download\n",
    "    def add_links(cleanjson,biotooljsonhit): \n",
    "        ## pool links\n",
    "        biotoollinks = []\n",
    "        if isinstance(biotooljsonhit['link'],list)==True:\n",
    "            for eachdict in biotooljsonhit['link']:\n",
    "                biotoollinks.append(eachdict)\n",
    "        if isinstance(biotooljsonhit['download'],list)==True:\n",
    "            for eachdict in biotooljsonhit['link']:\n",
    "                biotoollinks.append(eachdict)\n",
    "        if isinstance(biotooljsonhit['link'],dict)==True:\n",
    "            biotoollinks.append(biotooljsonhit['link'])\n",
    "        if isinstance(biotooljsonhit['download'],dict)==True:\n",
    "            biotoollinks.append(biotooljsonhit['download'])\n",
    "\n",
    "        ## Parse links\n",
    "        if len(biotoollinks) > 0:\n",
    "            codeRepository = []\n",
    "            discussionUrl = []\n",
    "            downloadUrl = []\n",
    "            for eachitem in biotoollinks:\n",
    "                if isinstance(eachitem,dict)==True and eachitem['type']!=None:\n",
    "                    if ('Repository' or 'repository') in eachitem['type']:\n",
    "                        codeRepository.append(eachitem['url'])\n",
    "                    if ('Issue' or 'issue') in eachitem['type']:\n",
    "                        discussionUrl.append(eachitem['url'])\n",
    "                    if ('file' in eachitem['type']):\n",
    "                        downloadUrl.append(eachitem['url'])\n",
    "                    if ('note' in eachitem.keys()) and (eachitem['note']!=None) and ('image' in eachitem['note']):\n",
    "                        downloadUrl.append(eachitem['url'])\n",
    "            if len(codeRepository)>0:\n",
    "                cleanjson['codeRepository'] = list(set(codeRepository))\n",
    "            if len(discussionUrl)>0:\n",
    "                cleanjson['discussionUrl'] = list(set(discussionUrl))\n",
    "            if len(downloadUrl)>0:\n",
    "                cleanjson['downloadUrl'] = list(set(downloadUrl))\n",
    "        return cleanjson\n",
    "    \n",
    "    ## parse softwarehelp from documentation\n",
    "    def add_softwarehelp(cleanjson,biotooljsonhit):\n",
    "        if len(biotooljsonhit['documentation'])>0:\n",
    "            if isinstance(biotooljsonhit['documentation'],list)==True:\n",
    "                cleanjson['softwareHelp'] = [x['url'] for x in biotooljsonhit['documentation']]\n",
    "            if isinstance(biotooljsonhit['documentation'],dict)==True:\n",
    "                cleanjson['softwareHelp'] = [biotooljsonhit['documentation']]\n",
    "        return cleanjson\n",
    "\n",
    "    ## parse authors from credit\n",
    "    def add_author(cleanjson,biotooljsonhit):\n",
    "        authorlist = []\n",
    "        for eachhit in biotooljsonhit['credit']:\n",
    "            newdict = copy.deepcopy(eachhit)\n",
    "            if newdict['typeEntity'] == 'outbreak:Person':\n",
    "                newdict['@type'] = 'outbreak:Person'\n",
    "            elif newdict['typeEntity'] == None: \n",
    "                if newdict['orcidid']!= None:\n",
    "                    newdict['@type'] = 'outbreak:Person'\n",
    "                else:\n",
    "                    newdict['@type'] = 'dct:Agent'\n",
    "            else:\n",
    "                newdict['@type'] = 'outbreak:Organization'\n",
    "            newdict['identifier'] = newdict.pop('orcidid',None)\n",
    "            newdict['role'] = newdict.pop('typeRole',None)\n",
    "            newdict.pop('gridid',None)\n",
    "            newdict.pop('rorid',None)\n",
    "            newdict.pop('fundrefid',None)\n",
    "            newdict.pop('note',None)\n",
    "            newdict.pop('typeEntity',None)\n",
    "            for key, value in dict(newdict).items():\n",
    "                if (value is None) or (len(value)==0):\n",
    "                    del newdict[key]\n",
    "            authorlist.append(newdict)\n",
    "        cleanjson['author'] = authorlist\n",
    "        return cleanjson\n",
    "    \n",
    "    ## parse citedBy from 'publication'\n",
    "    def add_citations(cleanjson,biotooljsonhit):\n",
    "        citedby = []\n",
    "        for eachpub in biotooljsonhit['publication']:\n",
    "            tmppub = copy.deepcopy(eachpub)\n",
    "            tmppub['@type'] = 'outbreak:Publication'\n",
    "            try:\n",
    "                tmppub['name'] = tmppub['metadata']['title']\n",
    "            except:\n",
    "                tmppub['name'] = None\n",
    "            tmppub.pop('type')\n",
    "            tmppub.pop('version')\n",
    "            tmppub.pop('note')\n",
    "            tmppub.pop('metadata')\n",
    "            for key, value in dict(tmppub).items():\n",
    "                if (value is None) or (len(value)==0):\n",
    "                    del tmppub[key]\n",
    "            citedby.append(tmppub)\n",
    "        cleanjson['citedBy'] = citedby\n",
    "        \n",
    "        return cleanjson\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5f92d7cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Fetch the json docs from the API\n",
    "def download_jsondocs():\n",
    "    jsondoclist = []\n",
    "    biotoolsapiurl = 'https://bio.tools/api/t'\n",
    "    payloads = {'format': 'json', 'domain': 'covid-19','page':'1'}\n",
    "    r = requests.get(biotoolsapiurl, params=payloads, timeout=5).json()\n",
    "    count = r['count']\n",
    "    list_num = len(r['list'])\n",
    "    total_pages = math.ceil(count/list_num) \n",
    "    i=1\n",
    "    while i < total_pages+1:\n",
    "        payloads = {'format': 'json', 'domain': 'covid-19','page':i}\n",
    "        r = requests.get(biotoolsapiurl, params=payloads, timeout=20).json()\n",
    "        time.sleep(1)\n",
    "        jsondoclist.extend(r['list'])\n",
    "        i=i+1\n",
    "    return jsondoclist\n",
    "\n",
    "    \n",
    "## Transform the json docs\n",
    "def transform_json(jsondoclist):\n",
    "    for i in range(len(jsondoclist)):\n",
    "        biotooljsonhit = jsondoclist[i]\n",
    "        cleanjson = cleandoc.add_basic_info(biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_app_sub_cat(cleanjson,biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_features(cleanjson,biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_links(cleanjson,biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_softwarehelp(cleanjson,biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_author(cleanjson,biotooljsonhit)\n",
    "        cleanjson = cleandoc.add_citations(cleanjson,biotooljsonhit)    \n",
    "        yield cleanNullTerms(cleanjson)\n",
    "        \n",
    "\n",
    "def load_annotations():\n",
    "    jsondoclist = download_jsondocs()\n",
    "    doclist = transform_json(jsondoclist)\n",
    "    yield from doclist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a53be9",
   "metadata": {},
   "source": [
    "### Check Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5c139ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "206\n",
      "21\n"
     ]
    }
   ],
   "source": [
    "biotoolsapiurl = 'https://bio.tools/api/t'\n",
    "payloads = {'format': 'json', 'domain': 'covid-19','page':'1'}\n",
    "#payloads = {'format': 'json', 'q': 'COVID-19','page':'1'}\n",
    "r = requests.get(biotoolsapiurl, params=payloads, timeout=5).json()\n",
    "count = r['count']\n",
    "list_num = len(r['list'])\n",
    "total_pages = math.ceil(count/list_num)\n",
    "print(list_num)\n",
    "print(count)\n",
    "print(total_pages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e81d46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "i=1\n",
    "while i < total_pages+1:\n",
    "    payloads = {'format': 'json', 'domain': 'covid-19','page':i}\n",
    "    r = requests.get(biotoolsapiurl, params=payloads, timeout=20).json()\n",
    "    time.sleep(1)\n",
    "    i=i+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd74adc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'name': 'Bridget P Bannerman', 'email': 'bpc28@cam.ac.uk', 'url': None, 'orcidid': 'https://orcid.org/0000-0002-5746-8283', 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': 'Person', 'typeRole': [], 'note': None}, {'name': 'R Andres Floto', 'email': 'arf27@cam.ac.uk', 'url': None, 'orcidid': 'https://orcid.org/0000-0002-2188-5659', 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': 'Person', 'typeRole': [], 'note': None}, {'name': 'Jorge Júlvez', 'email': None, 'url': None, 'orcidid': 'https://orcid.org/0000-0002-7093-228X', 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': None, 'typeRole': [], 'note': None}, {'name': 'Alexandru Oarga', 'email': None, 'url': None, 'orcidid': 'https://orcid.org/0000-0002-7271-733X', 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': None, 'typeRole': [], 'note': None}, {'name': 'Tom L Blundell', 'email': None, 'url': None, 'orcidid': None, 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': None, 'typeRole': [], 'note': None}, {'name': 'Pablo Moreno', 'email': None, 'url': None, 'orcidid': None, 'gridid': None, 'rorid': None, 'fundrefid': None, 'typeEntity': None, 'typeRole': [], 'note': None}]\n"
     ]
    }
   ],
   "source": [
    "print(r['list'][4]['credit'])\n",
    "biotooljsonhit = r['list'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fcdaeba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(biotooljsonhit.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28d20fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## parse doi from 'otherID' (if available)\n",
    "biotooljsonhit['otherID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b33d054b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'DisGeNET', 'description': 'DisGeNET is a discovery platform containing one of the largest publicly available collections of genes and variants associated to human diseases. DisGeNET integrates data from expert curated repositories, GWAS catalogues, animal models and the scientific literature. DisGeNET data are homogeneously annotated with controlled vocabularies and community-driven ontologies. Additionally, several original metrics are provided to assist the prioritization of genotype–phenotype relationships.\\n\\nThe current version of DisGeNET (v7.0) contains 1,134,942 gene-disease associations (GDAs), between 21,671 genes and 30,170 diseases, disorders, traits, and clinical or abnormal human phenotypes, and 369,554 variant-disease associations (VDAs), between 194,515 variants and 14,155 diseases, traits, and phenotypes.', 'identifier': 'disgenet', 'url': 'http://www.disgenet.org/', 'softwareVersion': ['Database version 7.0', 'Platform v1.1.0', 'REST API v1.0.0', 'DisGeNET RDF release 7.0', 'disgenet2r R package Version: 0.0.9', 'DisGeNET Cytoscape App Version 7.0.x'], 'applicationCategory': ['Web application', 'SPARQL endpoint', 'Plug-in', 'Web API', 'Database portal'], 'license': 'CC-BY-NC-SA-4.0', 'conditionsOfAccess': 'Open access (with restrictions)', 'dateModified': '2022-06-21', 'dateCreated': '2016-04-06', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0602', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0602', 'url': 'http://edamontology.org/topic_0602', 'name': 'Molecular interactions, pathways and networks'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3344', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3344', 'url': 'http://edamontology.org/topic_3344', 'name': 'Biomedical science'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0199', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0199', 'url': 'http://edamontology.org/topic_0199', 'name': 'Genetic variation'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0218', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0218', 'url': 'http://edamontology.org/topic_0218', 'name': 'Natural language processing'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0089', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0089', 'url': 'http://edamontology.org/topic_0089', 'name': 'Ontology and terminology'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0091', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0091', 'url': 'http://edamontology.org/topic_0091', 'name': 'Bioinformatics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3366', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3366', 'url': 'http://edamontology.org/topic_3366', 'name': 'Data integration and warehousing'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3345', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3345', 'url': 'http://edamontology.org/topic_3345', 'name': 'Data identity and mapping'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3342', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3342', 'url': 'http://edamontology.org/topic_3342', 'name': 'Translational medicine'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0634', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0634', 'url': 'http://edamontology.org/topic_0634', 'name': 'Pathology'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0625', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0625', 'url': 'http://edamontology.org/topic_0625', 'name': 'Genotype and phenotype'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3063', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3063', 'url': 'http://edamontology.org/topic_3063', 'name': 'Medical informatics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3325', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3325', 'url': 'http://edamontology.org/topic_3325', 'name': 'Rare diseases'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_2421', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_2421', 'url': 'http://edamontology.org/operation_2421', 'name': 'Database search'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3226', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3226', 'url': 'http://edamontology.org/operation_3226', 'name': 'Variant prioritisation'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3083', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3083', 'url': 'http://edamontology.org/operation_3083', 'name': 'Pathway or network visualisation'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3436', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3436', 'url': 'http://edamontology.org/operation_3436', 'name': 'Aggregation'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_2422', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_2422', 'url': 'http://edamontology.org/operation_2422', 'name': 'Data retrieval'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1027', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1027', 'url': 'http://edamontology.org/data_1027', 'name': 'Gene ID (NCBI)'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3021', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3021', 'url': 'http://edamontology.org/data_3021', 'name': 'UniProt accession'}, {'@type': 'schema:DefinedTerm', '@id': 'data_1106', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1106', 'url': 'http://edamontology.org/data_1106', 'name': 'dbSNP ID'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2330', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2330', 'url': 'http://edamontology.org/format_2330', 'name': 'Textual format'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_2080', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_2080', 'url': 'http://edamontology.org/data_2080', 'name': 'Database search results'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_3475', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3475', 'url': 'http://edamontology.org/format_3475', 'name': 'TSV'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2376', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2376', 'url': 'http://edamontology.org/format_2376', 'name': 'RDF format'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3464', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3464', 'url': 'http://edamontology.org/format_3464', 'name': 'JSON'}]}], 'codeRepository': ['http://www.disgenet.org/'], 'softwareHelp': ['http://www.disgenet.org/dbinfo', 'https://www.disgenet.org/api/'], 'author': [{'name': 'Ferran Sanz', 'email': 'ferran.sanz@upf.edu', '@type': 'Person', 'identifier': 'http://orcid.org/0000-0002-7534-7661', 'role': ['Contributor']}, {'name': 'Janet Piñero', 'email': 'janet.pinero@upf.edu', '@type': 'Person', 'identifier': 'http://orcid.org/0000-0003-1244-7654', 'role': ['Contributor', 'Developer', 'Maintainer', 'Support']}, {'name': 'Laura I. Furlong', 'email': 'laura.furlong@upf.edu', 'url': 'http://grib.imim.es/about-us/laurai-furlong/', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-9383-528X', 'role': ['Primary contact']}, {'name': 'Josep Sauch Pitchard', 'email': 'jsauch@imim.es', '@type': 'Person', 'role': ['Developer']}, {'name': 'Juan Manuel Ramirez Anguita', '@type': 'Person', 'role': ['Developer']}, {'name': 'Francesco Ronzano', 'email': 'francesco.ronzano@upf.edu', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0001-5037-9061', 'role': ['Contributor']}], 'citedBy': [{'doi': '10.1093/database/bav028', '@type': 'Publication', 'name': 'DisGeNET: A discovery platform for the dynamical exploration of human diseases and their genes'}, {'doi': '10.1093/nar/gkw943', '@type': 'Publication', 'name': 'DisGeNET: A comprehensive platform integrating information on human disease-associated genes and variants'}, {'doi': '10.1093/bioinformatics/btq538', '@type': 'Publication', 'name': 'DisGeNET: A Cytoscape plugin to visualize, integrate, search and analyze gene-disease networks'}, {'doi': '10.1093/bioinformatics/btw214', '@type': 'Publication', 'name': 'DisGeNET-RDF: Harnessing the innovative power of the Semantic Web to explore the genetic basis of diseases'}, {'doi': '10.1371/journal.pone.0020284', '@type': 'Publication', 'name': 'Gene-disease network analysis reveals functional modules in mendelian, complex and environmental diseases'}, {'doi': '10.1093/nar/gkw943', '@type': 'Publication', 'name': 'DisGeNET: A comprehensive platform integrating information on human disease-associated genes and variants'}, {'doi': '10.1093/nar/gkz1021', '@type': 'Publication', 'name': 'The DisGeNET knowledge platform for disease genomics: 2019 update'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'V-pipe', 'description': 'Bioinformatics pipeline for the analysis of next-generation sequencing data derived from intra-host viral populations.', 'identifier': 'v-pipe', 'url': 'https://cbg-ethz.github.io/V-pipe/', 'softwareVersion': ['v2.99.2', 'v2.99.1', 'v2.99.0', 'v2.0', 'v1.0'], 'applicationCategory': ['Command-line tool', 'Workflow'], 'license': 'Apache-2.0', 'programmingLanguage': ['Shell', 'Python'], 'conditionsOfAccess': 'Open access', 'dateModified': '2022-03-31', 'dateCreated': '2018-02-02', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0622', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0622', 'url': 'http://edamontology.org/topic_0622', 'name': 'Genomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3056', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3056', 'url': 'http://edamontology.org/topic_3056', 'name': 'Population genetics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0769', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0769', 'url': 'http://edamontology.org/topic_0769', 'name': 'Workflows'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0781', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0781', 'url': 'http://edamontology.org/topic_0781', 'name': 'Virology'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3168', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3168', 'url': 'http://edamontology.org/topic_3168', 'name': 'Sequencing'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_3219', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3219', 'url': 'http://edamontology.org/operation_3219', 'name': 'Read pre-processing'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0292', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0292', 'url': 'http://edamontology.org/operation_0292', 'name': 'Sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3197', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3197', 'url': 'http://edamontology.org/operation_3197', 'name': 'Genetic variation analysis'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_2975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_2975', 'url': 'http://edamontology.org/data_2975', 'name': 'Nucleic acid sequence (raw)'}], 'defaultValue': []}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0920', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0920', 'url': 'http://edamontology.org/data_0920', 'name': 'Genotype/phenotype report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}]}], 'codeRepository': ['https://github.com/cbg-ethz/V-pipe'], 'softwareHelp': ['https://cbg-ethz.github.io/V-pipe/usage/', 'https://github.com/cbg-ethz/V-pipe/blob/master/README.md', 'https://github.com/cbg-ethz/V-pipe/blob/master/config/README.md', 'https://github.com/cbg-ethz/V-pipe/wiki/benchmark', 'https://github.com/cbg-ethz/V-pipe/wiki/snakemake', 'https://youtu.be/qHEUVJZsgE4', 'https://cbg-ethz.github.io/V-pipe/tutorial/sars-cov2/', 'https://youtu.be/pIby1UooK94'], 'author': [{'name': 'Susana Posada Céspedes', 'email': 'susana.posada@bsse.ethz.ch', 'url': 'https://github.com/sposadac', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-7459-8186', 'role': ['Developer']}, {'name': 'David Seifert', 'email': 'david.seifert@bsse.ethz.ch', 'url': 'https://github.com/SoapZA', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0003-4739-5110', 'role': ['Developer']}, {'name': 'Niko Beerenwinkel', 'email': 'niko.beerenwinkel@bsse.ethz.ch', 'url': 'https://www.bsse.ethz.ch/cbg/group/people/person-detail.html?persid=149417', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-0573-6119', 'role': ['Contributor']}, {'name': 'Ivan Topolsky', 'email': 'ivan.topolsky@sib.swiss', 'url': 'https://www.sib.swiss/directory?cn=ivan%20topolsky', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-7561-0810', 'role': ['Maintainer']}, {'name': 'Computational Biology Group', 'url': 'https://www.bsse.ethz.ch/cbg', '@type': 'Organization'}, {'name': 'D-BSSE', 'url': 'https://www.bsse.ethz.ch/', '@type': 'Organization'}, {'name': 'ETHZ', 'url': 'https://www.ethz.ch/', '@type': 'Organization'}, {'name': 'SIB Swiss Institute of Bioinformatics', 'url': 'https://www.sib.swiss/', '@type': 'Organization'}, {'name': 'V-pipe Dev Team', 'email': 'v-pipe@bsse.ethz.ch', '@type': 'Person', 'role': ['Primary contact']}], 'citedBy': [{'doi': '10.1093/bioinformatics/btab015', 'pmid': '33471068', '@type': 'Publication', 'name': 'V-pipe: A computational pipeline for assessing viral genetic diversity from high-throughput data'}, {'doi': '10.20944/preprints202005.0376.v1', '@type': 'Publication'}, {'doi': '10.1016/j.virusres.2016.09.016', 'pmid': '27693290', '@type': 'Publication', 'name': 'Recent advances in inferring viral diversity from high-throughput sequencing data'}, {'doi': '10.3390/v10050256', 'pmid': '29757994', 'pmcid': 'PMC5977249', '@type': 'Publication', 'name': 'Bioinformatics meets virology: The European virus bioinformatics center’s second annual meeting'}, {'doi': '10.1093/infdis/jiz094', 'pmid': '30835266', '@type': 'Publication', 'name': 'Viral Diversity Based on Next-Generation Sequencing of HIV-1 Provides Precise Estimates of Infection Recency and Time since Infection'}, {'doi': '10.2807/1560-7917.es.2020.25.32.2001410', '@type': 'Publication', 'name': 'Geographical and temporal distribution of SARS-CoV-2 clades in the WHO European Region, January to June 2020'}, {'doi': '10.1101/2020.10.12.335919', '@type': 'Publication'}, {'doi': '10.1101/2020.10.14.20212621', '@type': 'Publication'}, {'doi': '10.1101/2021.01.08.21249379', '@type': 'Publication'}, {'doi': '10.1016/j.epidem.2021.100480', 'pmid': '34488035', '@type': 'Publication', 'name': 'Quantification of the spread of SARS-CoV-2 variant B.1.1.7 in Switzerland'}, {'doi': '10.3390/v13030496', 'pmid': '33802899', 'pmcid': 'PMC8002591', '@type': 'Publication', 'name': 'Detection and genome sequencing of sars-cov-2 in a domestic cat with respiratory signs in Switzerland'}, {'doi': '10.1016/j.coviro.2021.06.002', 'pmid': '34153841', '@type': 'Publication', 'name': 'Quantitative measures of within-host viral genetic diversity'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'impuSARS', 'description': 'impuSARS allows the imputation of viral whole genome sequences from partially sequenced samples. Additionally, impuSARS provides the lineage associated to the imputed sequence. impuSARS have been validated with a reference of SARS-CoV-2 sequences.', 'identifier': 'impusars', 'url': 'https://github.com/babelomics/impuSARS', 'softwareVersion': ['v1.0', 'v2.0', 'v3.0', 'v3.1'], 'applicationCategory': ['Command-line tool'], 'license': 'MIT', 'programmingLanguage': ['Python', 'Bash'], 'conditionsOfAccess': 'Open access', 'dateModified': '2022-03-11', 'dateCreated': '2021-09-10', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_3307', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3307', 'url': 'http://edamontology.org/topic_3307', 'name': 'Computational biology'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0080', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0080', 'url': 'http://edamontology.org/topic_0080', 'name': 'Sequence analysis'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0091', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0091', 'url': 'http://edamontology.org/topic_0091', 'name': 'Bioinformatics'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_3219', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3219', 'url': 'http://edamontology.org/operation_3219', 'name': 'Read pre-processing'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0292', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0292', 'url': 'http://edamontology.org/operation_0292', 'name': 'Sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3197', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3197', 'url': 'http://edamontology.org/operation_3197', 'name': 'Genetic variation analysis'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_2975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_2975', 'url': 'http://edamontology.org/data_2975', 'name': 'Nucleic acid sequence (raw)'}], 'defaultValue': []}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0920', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0920', 'url': 'http://edamontology.org/data_0920', 'name': 'Genotype/phenotype report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}]}], 'codeRepository': ['https://github.com/cbg-ethz/V-pipe'], 'softwareHelp': ['https://cbg-ethz.github.io/V-pipe/usage/', 'https://github.com/cbg-ethz/V-pipe/blob/master/README.md', 'https://github.com/cbg-ethz/V-pipe/blob/master/config/README.md', 'https://github.com/cbg-ethz/V-pipe/wiki/benchmark', 'https://github.com/cbg-ethz/V-pipe/wiki/snakemake', 'https://youtu.be/qHEUVJZsgE4', 'https://cbg-ethz.github.io/V-pipe/tutorial/sars-cov2/', 'https://youtu.be/pIby1UooK94']}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'nf-core-viralrecon', 'description': 'nfcore/viralrecon is a bioinformatics analysis pipeline used to perform assembly and intra-host/low-frequency variant calling for viral samples. The pipeline supports short-read Illumina sequencing data from both shotgun (e.g. sequencing directly from clinical samples) and enrichment-based library preparation methods (e.g. amplicon-based: ARTIC SARS-CoV-2 enrichment protocol; or probe-capture-based).', 'identifier': 'nf-core-viralrecon', 'url': 'https://nf-co.re/viralrecon', 'softwareVersion': ['1.0.0', '1.1.0', '2.0', '2.1', '2.2', '2.3', '2.3.1', '2.4', '2.4.1'], 'applicationCategory': ['Command-line tool', 'Workflow'], 'license': 'MIT', 'programmingLanguage': ['Groovy'], 'conditionsOfAccess': 'Open access', 'dateModified': '2022-03-11', 'dateCreated': '2021-01-13', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0622', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0622', 'url': 'http://edamontology.org/topic_0622', 'name': 'Genomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0781', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0781', 'url': 'http://edamontology.org/topic_0781', 'name': 'Virology'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3796', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3796', 'url': 'http://edamontology.org/topic_3796', 'name': 'Population genomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0199', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0199', 'url': 'http://edamontology.org/topic_0199', 'name': 'Genetic variation'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3168', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3168', 'url': 'http://edamontology.org/topic_3168', 'name': 'Sequencing'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0102', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0102', 'url': 'http://edamontology.org/topic_0102', 'name': 'Mapping'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0769', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0769', 'url': 'http://edamontology.org/topic_0769', 'name': 'Workflows'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_0292', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0292', 'url': 'http://edamontology.org/operation_0292', 'name': 'Sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3197', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3197', 'url': 'http://edamontology.org/operation_3197', 'name': 'Genetic variation analysis'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0525', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0525', 'url': 'http://edamontology.org/operation_0525', 'name': 'Genome assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0346', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0346', 'url': 'http://edamontology.org/operation_0346', 'name': 'Sequence similarity search'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3227', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3227', 'url': 'http://edamontology.org/operation_3227', 'name': 'Variant calling'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3731', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3731', 'url': 'http://edamontology.org/operation_3731', 'name': 'Sample comparison'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_3494', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3494', 'url': 'http://edamontology.org/data_3494', 'name': 'DNA sequence'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_1930', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1930', 'url': 'http://edamontology.org/format_1930', 'name': 'FASTQ'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/nf-core/viralrecon'], 'softwareHelp': ['https://nf-co.re/viralrecon', 'https://nf-co.re/viralrecon/usage', 'https://github.com/nf-core/viralrecon/blob/master/LICENSE'], 'author': [{'name': 'Sarai Varona', 'email': 's.varona@isciii.es', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-2245-210X', 'role': ['Primary contact', 'Developer', 'Maintainer']}, {'name': 'Sara Monzón', 'email': 'smonzon@isciii.es', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-0740-6324', 'role': ['Primary contact', 'Developer', 'Maintainer']}, {'name': 'Harshil Patel', 'email': 'Harshil.Patel@crick.ac.uk', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0003-2707-7940', 'role': ['Primary contact', 'Developer', 'Maintainer']}, {'name': 'Philip A. Ewels', 'email': 'phil.ewels@scilifelab.se', 'url': 'https://phil.ewels.co.uk/', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0003-4101-2502', 'role': ['Primary contact']}, {'name': 'Maxime U. Garcia', 'email': 'maxime.garcia@scilifelab.se', 'url': 'https://maxulysse.github.io/', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0003-2827-9261', 'role': ['Primary contact']}, {'name': 'Jose Espinosa-Carrasco', 'email': 'joseantonio.espinosa@crg.eu', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-1541-042X', 'role': ['Primary contact']}, {'name': 'nf-core', 'url': 'https://nf-co.re', '@type': 'Organization'}], 'citedBy': [{'doi': '10.1038/s41587-020-0439-x', 'pmid': '32055031', '@type': 'Publication', 'name': 'The nf-core framework for community-curated bioinformatics pipelines'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'ifindCPcli', 'description': 'findCPcli is a command line python-tool for the computation of chokepoint reactions in genome-scale metabolic models. The main purpose of the tool is to compute chokepoints by taking into account both the topology and the dynamic information of the network. In addition to the computation of chokepoints, findCPcli can compute and remove dead-end metabolites, find essential reactions and update the flux bounds of the reactions according to the results of Flux Variability Analysis.\\n\\nfindCPcli takes as input an SBML files of genome-scale models, and provides as output a spreadsheet file with the results of the chokepoint computation.', 'identifier': 'ifindcpcli', 'url': 'https://github.com/findCP/findCPcli', 'applicationCategory': ['Command-line tool'], 'license': 'GPL-3.0', 'programmingLanguage': ['Python', 'Shell'], 'conditionsOfAccess': 'Open access', 'dateModified': '2022-01-18', 'dateCreated': '2022-01-18', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0154', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0154', 'url': 'http://edamontology.org/topic_0154', 'name': 'Small molecules'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3407', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3407', 'url': 'http://edamontology.org/topic_3407', 'name': 'Endocrinology and metabolism'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3172', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3172', 'url': 'http://edamontology.org/topic_3172', 'name': 'Metabolomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0602', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0602', 'url': 'http://edamontology.org/topic_0602', 'name': 'Molecular interactions, pathways and networks'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0821', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0821', 'url': 'http://edamontology.org/topic_0821', 'name': 'Enzymes'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_3660', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3660', 'url': 'http://edamontology.org/operation_3660', 'name': 'Metabolic network modelling'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_4009', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_4009', 'url': 'http://edamontology.org/operation_4009', 'name': 'Small molecule design'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3929', 'url': 'http://edamontology.org/operation_3929', 'name': 'Metabolic pathway prediction'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/nf-core/viralrecon'], 'softwareHelp': ['https://findcpcli.readthedocs.io/en/latest/'], 'author': [{'name': 'Bridget P Bannerman', 'email': 'bpc28@cam.ac.uk', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-5746-8283'}, {'name': 'R Andres Floto', 'email': 'arf27@cam.ac.uk', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-2188-5659'}, {'name': 'Jorge Júlvez', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-7093-228X'}, {'name': 'Alexandru Oarga', '@type': 'Person', 'identifier': 'https://orcid.org/0000-0002-7271-733X'}, {'name': 'Tom L Blundell', '@type': 'Organization'}, {'name': 'Pablo Moreno', '@type': 'Organization'}], 'citedBy': [{'doi': '10.26508/LSA.202000954', 'pmid': '34353886', 'pmcid': 'PMC8343166', '@type': 'Publication', 'name': 'Integrated human/SARS-CoV-2 metabolic models present novel treatment strategies against COVID-19'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'Llamanade', 'description': 'An open-source computational pipeline for robust nanobody humanization.', 'identifier': 'llamanade', 'url': 'https://github.com/sangzhe/Llamanade', 'applicationCategory': ['Script'], 'license': 'Not licensed', 'programmingLanguage': ['Python', 'Shell'], 'conditionsOfAccess': 'Open access', 'dateModified': '2022-01-14', 'dateCreated': '2022-01-14', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_3379', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3379', 'url': 'http://edamontology.org/topic_3379', 'name': 'Preclinical and clinical studies'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_2830', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_2830', 'url': 'http://edamontology.org/topic_2830', 'name': 'Immunoproteins and antigens'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3168', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3168', 'url': 'http://edamontology.org/topic_3168', 'name': 'Sequencing'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0769', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0769', 'url': 'http://edamontology.org/topic_0769', 'name': 'Workflows'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3398', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3398', 'url': 'http://edamontology.org/topic_3398', 'name': 'Bioengineering'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_0480', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0480', 'url': 'http://edamontology.org/operation_0480', 'name': 'Side chain modelling'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_2950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_2950', 'url': 'http://edamontology.org/operation_2950', 'name': 'Residue distance calculation'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0337', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0337', 'url': 'http://edamontology.org/operation_0337', 'name': 'Visualisation'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/sangzhe/Llamanade-scripts'], 'softwareHelp': ['https://findcpcli.readthedocs.io/en/latest/'], 'author': [{'name': 'Yi Shi', 'email': 'yi.shi@pitt.edu', '@type': 'Person'}, {'name': 'Zhe Sang', '@type': 'Organization'}, {'name': 'Yufei Xiang', '@type': 'Organization'}, {'name': 'Ivet Bahar', '@type': 'Organization'}], 'citedBy': [{'doi': '10.1101/2021.08.03.455003', 'pmid': '34373858', 'pmcid': 'PMC8351782', '@type': 'Publication'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'CovidPhy', 'description': 'Web interface that can process SARS-CoV-2 genome sequences in plain fasta text format or provided through identity codes from the Global Initiative on Sharing Avian Influenza Data (GISAID) or GenBank.', 'identifier': 'covidphy', 'url': 'http://covidphy.eu', 'applicationCategory': ['Web application'], 'programmingLanguage': ['JavaScript'], 'conditionsOfAccess': 'Open access', 'dateModified': '2021-12-31', 'dateCreated': '2021-12-31', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0084', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0084', 'url': 'http://edamontology.org/topic_0084', 'name': 'Phylogeny'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0099', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0099', 'url': 'http://edamontology.org/topic_0099', 'name': 'RNA'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0080', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0080', 'url': 'http://edamontology.org/topic_0080', 'name': 'Sequence analysis'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0199', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0199', 'url': 'http://edamontology.org/topic_0199', 'name': 'Genetic variation'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_3436', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3436', 'url': 'http://edamontology.org/operation_3436', 'name': 'Aggregation'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_2421', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_2421', 'url': 'http://edamontology.org/operation_2421', 'name': 'Database search'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0548', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0548', 'url': 'http://edamontology.org/operation_0548', 'name': 'Phylogenetic inference (quartet methods)'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/sangzhe/Llamanade-scripts'], 'softwareHelp': ['https://findcpcli.readthedocs.io/en/latest/'], 'author': [{'name': 'Antonio Salas', '@type': 'Person'}], 'citedBy': [{'doi': '10.1016/J.ENVRES.2021.111909', 'pmid': '34419470', 'pmcid': 'PMC8376833', '@type': 'Publication', 'name': 'CovidPhy: A tool for phylogeographic analysis of SARS-CoV-2 variation'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'COVID-ONE-humoral immune', 'description': 'The One-stop Database for COVID-19-specific Antibody Responses and Clinical Parameters.', 'identifier': 'pub2tools2021__covid-one-humoral-immune', 'url': 'http://www.COVID-ONE.cn', 'applicationCategory': ['Database portal'], 'conditionsOfAccess': 'Open access', 'dateModified': '2021-12-31', 'dateCreated': '2021-12-31', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_3518', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3518', 'url': 'http://edamontology.org/topic_3518', 'name': 'Microarray experiment'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0121', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0121', 'url': 'http://edamontology.org/topic_0121', 'name': 'Proteomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0154', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0154', 'url': 'http://edamontology.org/topic_0154', 'name': 'Small molecules'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0203', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0203', 'url': 'http://edamontology.org/topic_0203', 'name': 'Gene expression'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0080', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0080', 'url': 'http://edamontology.org/topic_0080', 'name': 'Sequence analysis'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_0252', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0252', 'url': 'http://edamontology.org/operation_0252', 'name': 'Peptide immunogenicity prediction'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0314', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0314', 'url': 'http://edamontology.org/operation_0314', 'name': 'Gene expression profiling'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3431', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3431', 'url': 'http://edamontology.org/operation_3431', 'name': 'Deposition'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/sangzhe/Llamanade-scripts'], 'softwareHelp': ['https://findcpcli.readthedocs.io/en/latest/'], 'author': [{'name': 'Feng Wang', 'email': 'taosc@sjtu.edu.cn', '@type': 'Person'}, {'name': 'Xionglin Fan', 'email': 'xlfan@hust.edu.cn', '@type': 'Person'}, {'name': 'Sheng-ce Tao', 'email': 'fengwang@tjh.tjmu.edu.cn', '@type': 'Person'}], 'citedBy': [{'doi': '10.1101/2021.07.29.21261312', '@type': 'Publication'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'coronaSPAdes', 'description': 'CoronaSPAdes – a novel assembler for RNA viral species recovery in general and coronaviruses in particular.', 'identifier': 'coronaspades', 'url': 'http://cab.spbu.ru/software/spades', 'applicationCategory': ['Library'], 'programmingLanguage': ['Python'], 'conditionsOfAccess': 'Open access', 'dateModified': '2021-12-19', 'dateCreated': '2021-12-19', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_3308', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3308', 'url': 'http://edamontology.org/topic_3308', 'name': 'Transcriptomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0196', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0196', 'url': 'http://edamontology.org/topic_0196', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3174', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3174', 'url': 'http://edamontology.org/topic_3174', 'name': 'Metagenomics'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_0160', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0160', 'url': 'http://edamontology.org/topic_0160', 'name': 'Sequence sites, features and motifs'}, {'@type': 'schema:DefinedTerm', '@id': 'topic_3297', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_3297', 'url': 'http://edamontology.org/topic_3297', 'name': 'Biotechnology'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_0524', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0524', 'url': 'http://edamontology.org/operation_0524', 'name': 'De-novo assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3258', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3258', 'url': 'http://edamontology.org/operation_3258', 'name': 'Transcriptome assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0525', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0525', 'url': 'http://edamontology.org/operation_0525', 'name': 'Genome assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3216', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3216', 'url': 'http://edamontology.org/operation_3216', 'name': 'Scaffolding'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/ablab/spades'], 'softwareHelp': ['https://cab.spbu.ru/files/release3.15.3/manual.html'], 'author': [{'name': 'Anton Korobeynikov', 'email': 'a.korobeynikov@spbu.ru', '@type': 'Person'}], 'citedBy': [{'doi': '10.1093/BIOINFORMATICS/BTAB597', 'pmid': '34406356', '@type': 'Publication'}]}\n",
      "{'@type': 'outbreak:ComputationalTool', '@context': {'schema': 'http://schema.org/', 'outbreak': 'https://discovery.biothings.io/view/outbreak/'}, 'curatedBy': {'@type': 'outbreak:Organization', 'name': 'ELIXIR bio.tools', 'url': 'https://bio.tools/t?domain=covid-19', 'identifier': 'biotools', 'affiliation': [{'@type': 'outbreak:Organization', 'name': 'ELIXIR'}], 'curationDate': '2022-08-04'}, 'name': 'PHA4GE SARS-CoV-2 Contextual Data Specification', 'description': 'Software package as collection template and associated materials for SARS-CoV-2 metadata.\\nContains metadata collection template based on data standard designed to harmonize genomic surveillance contextual data (sample metadata, epidemiological/clinical/laboratory data, methods information) across datasets, laboratories and databases during pandemic, and beyond. Template contains standardized fields, terms and formats for structuring information, and is accompanied by field level reference guide which provides definitions, guidance and examples for populating fields. Fields have been prioritized using colour coding system as required (yellow), recommended (purple), and optional (white). Users can use as much or as little of specification as they wish, according to their needs, and need not share information unless permitted to do so. SOP and protocols for submitting to public repositories are also included.', 'identifier': 'pha4ge_sars-cov-2_contextual_data_specification', 'url': 'https://github.com/pha4ge/SARS-CoV-2-Contextual-Data-Specification', 'applicationCategory': ['Desktop application'], 'license': 'CC-BY-4.0', 'conditionsOfAccess': 'Open access', 'dateModified': '2021-12-08', 'dateCreated': '2021-12-08', 'applicationSubCategory': [{'@type': 'schema:DefinedTerm', '@id': 'topic_0089', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'topic_0089', 'url': 'http://edamontology.org/topic_0089', 'name': 'Ontology and terminology'}], 'featuresList': [{'@type': 'schema:DefinedTerm', '@id': 'operation_2409', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_2409', 'url': 'http://edamontology.org/operation_2409', 'name': 'Data handling'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_0335', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_0335', 'url': 'http://edamontology.org/operation_0335', 'name': 'Formatting'}, {'@type': 'schema:DefinedTerm', '@id': 'operation_3096', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'operation_3096', 'url': 'http://edamontology.org/operation_3096', 'name': 'Editing'}], 'input': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_0950', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0950', 'url': 'http://edamontology.org/data_0950', 'name': 'Mathematical model'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2585', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2585', 'url': 'http://edamontology.org/format_2585', 'name': 'SBML'}]}], 'output': [{'@type': 'bioschemastypes:FormalParameter', 'encodingFormat': [{'@type': 'schema:DefinedTerm', '@id': 'data_1383', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_1383', 'url': 'http://edamontology.org/data_1383', 'name': 'Nucleic acid sequence alignment'}, {'@type': 'schema:DefinedTerm', '@id': 'data_0925', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_0925', 'url': 'http://edamontology.org/data_0925', 'name': 'Sequence assembly'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3498', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3498', 'url': 'http://edamontology.org/data_3498', 'name': 'Sequence variations'}, {'@type': 'schema:DefinedTerm', '@id': 'data_3914', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'data_3914', 'url': 'http://edamontology.org/data_3914', 'name': 'Quality control report'}], 'defaultValue': [{'@type': 'schema:DefinedTerm', '@id': 'format_2572', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2572', 'url': 'http://edamontology.org/format_2572', 'name': 'BAM'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3327', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3327', 'url': 'http://edamontology.org/format_3327', 'name': 'BAI'}, {'@type': 'schema:DefinedTerm', '@id': 'format_1929', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_1929', 'url': 'http://edamontology.org/format_1929', 'name': 'FASTA'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3975', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3975', 'url': 'http://edamontology.org/format_3975', 'name': 'GFA 1'}, {'@type': 'schema:DefinedTerm', '@id': 'format_3016', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_3016', 'url': 'http://edamontology.org/format_3016', 'name': 'VCF'}, {'@type': 'schema:DefinedTerm', '@id': 'format_2331', 'inDefinedTermSet': 'http://edamontology.org/', 'termCode': 'format_2331', 'url': 'http://edamontology.org/format_2331', 'name': 'HTML'}]}], 'codeRepository': ['https://github.com/ablab/spades'], 'softwareHelp': ['https://cab.spbu.ru/files/release3.15.3/manual.html'], 'citedBy': [{'doi': '10.20944/preprints202008.0220.v1', '@type': 'Publication'}]}\n"
     ]
    }
   ],
   "source": [
    "cleanjson = transform_json(r['list'])\n",
    "for eachjson in cleanjson:\n",
    "    print(eachjson)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c121477",
   "metadata": {},
   "source": [
    "### Run time check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4de1db08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "206\n",
      "Wall time: 38.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "jsondoclist = download_jsondocs()\n",
    "print(len(jsondoclist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d201f919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "206\n",
      "Wall time: 1.77 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "doclist = transform_json(jsondoclist)\n",
    "\n",
    "i=0\n",
    "for doc in doclist:\n",
    "    i=i+1\n",
    "print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b9de828",
   "metadata": {},
   "source": [
    "### Integrated, and no longer needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891d5a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_version(cleanjson, biotooljsonhit):\n",
    "    try:\n",
    "        cleanjson['softwareVersion'] = biotooljsonhit['version']\n",
    "    except:\n",
    "        pass\n",
    "    return(cleanjson) \n",
    "   \n",
    "    \n",
    "def add_application_cat(cleanjson, biotooljsonhit):\n",
    "    try:\n",
    "        cleanjson['applicationCategory'] = biotooljsonhit['toolType']\n",
    "    except:\n",
    "        pass\n",
    "    return(cleanjson)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a04ab1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cleanjson = cleandoc.add_basic_info(biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_app_sub_cat(cleanjson,biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_features(cleanjson,biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_links(cleanjson,biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_softwarehelp(cleanjson,biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_author(cleanjson,biotooljsonhit)\n",
    "#print(cleanjson)\n",
    "cleanjson = cleandoc.add_citations(cleanjson,biotooljsonhit)\n",
    "print(cleanjson)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
